### ISSUE TITLE
Prevent user APK crashing b/c of bad/missing permission

### ISSUE THREAD

**Issue by mougino** [RFO-BASIC/Basic#220]
After a long session of PMs, debug, and tests with forum user Cassiope34, I was finally able to find the reason why [his APK was crashing](http://rfobasic.freeforums.org/zabble-kind-of-scrabble-with-dice-t4661.html).
Short answer: he removed the `READ_PHONE_STATE` permission that is needed by the BASIC! command `DEVICE` and when BASIC! runs this command w/o permission it does an ugly crash (with no information whatsoever, so very hard to debug).

It is not the first time users report a crash of BASIC! in their user APK and after investigation it is due to a missing permission.

But there's hope ;)

I stumbled upon this nice [stackoverflow](http://stackoverflow.com/questions/7203668/how-permission-can-be-checked-at-runtime-without-throwing-securityexception):
> How permission can be checked at runtime without throwing SecurityException

First answer gives a piece of code that we could adapt to check for existence of any permission in the user apk, and throw a clean error instead of an ugly crash:

```Java
private boolean checkPermission(String permission)
// e.g. checkPermission("WRITE_EXTERNAL_STORAGE")
{
    int res = getContext().checkCallingOrSelfPermission("android.permission." + permission);
    return (res == PackageManager.PERMISSION_GRANTED);            
}
```

`checkCallingOrSelfPermission()` was added in API level 1, so it's all good for us.

The only downsize is that we need to call checkPermission() a little everywhere, in all functions that need a specific permission.
Hopefully I have a chart that I use as an algorithm in Quick APK / BASIC! Compiler to auto-set the user APK permissions depending on the functions detected in his/her program.
So I'm willing to give it a go and propose something...

**Comment by mougino** [comment]
Here is the test program that I'm using to put in evidence the crashes in case of missing permission and fix them: http://laughton.com/basic/programs/other/check_perms/check_perms.bas

From this .bas I made 3 APKs:
* the first one based on current BASIC! repo, with all permissions (I let Quick APK auto-detect them)
* the second one based on current BASIC! repo with no permission except `WRITE_EXTERNAL_STORAGE`
* the third one based on my BASIC! attempt at fixing the crashes, also with no permission like APK 2

First APK is only to check the commands act normal and my test program is relevant.
It it the same to run this APK than to run the .bas directly in Editor mode (which has all the permissions).

Second APK shows me the test-cases (commands) where the crashes happen. There are 6 crashes:
* 1. `DEVICE` command w/o `READ_PHONE_STATE` permission crashes BASIC!
* 2. `GPS.*` commands w/o any of the `ACCESS_*_LOCATION` permissions crash BASIC!
* 4. `WAKELOCK` command w/o `WAKELOCK` permission crashes BASIC!
* 11. `BT.*` commands w/o `BLUETOOTH` nor `BLUETOOTH_ADMIN` permissions crash BASIC!
* 17. `PHONE.INFO` command w/o `READ_PHONE_STATE` crashes BASIC!
* 18. `MYPHONENUMBER` command w/o `READ_PHONE_STATE` crashes BASIC!

All the other commands cleanly handle the absence of permission.

Some commands already report that the needed permission is absent: that is the case for `VIBRATE`, `FTP.*` commands, `SMS.SEND`, `PHONE.CALL` (but this one is very ugly, showing (too technical) dumps of the failing Intent and Process), `WIFI.INFO`, and `HTTP.POST`.
However most of the times the message leaves room to user doubt: e.g. "Permission denied (missing INTERNET permission?)". I am for removing the question mark and harmonizing the message through all commands.

Some commands do not report missing permission but return an empty result: `GR.CAMERA` without the `CAMERA` permission does not crash but returns a null bitmap pointer ; `GRABURL` w/o the `INTERNET` permission returns an empty string and `SMS.RCV.*` commands w/o `RECEIVE_SMS` return nothing. `PHONE.RCV.*` w/o `READ_PHONE_STATE` returns an empty phone number. `SOCKET.*` without the `INTERNET` permission does not open the socket.
In these cases, I don't think we should return an empty result, because the user may think it comes from his program and try to debug for a loooong time. But if we threw a runtime error instead, he would know for certain his APK has 0% chance returning a valid result b/c it misses the needed permission...

Finally there were the surprises: I learned that on my phone STT and TTS don't need the `INTERNET` permission, there seems to be a local copy of Google files allowing STT/TTS to work offline... Should we still throw a runtime error if the user APK calls STT/TTS without the `INTERNET` permission ? (in case the phone doesn't have the database locally?)

**Comment by mougino** [comment]
Update: STT and TTS **do** work online and need a connection to work, just the connection is out of the app: even if the app has no `INTERNET` permission, calling STT and TTS opens a different window (process) that uses the network to work.
So I did not put any permission check in the `TTS.*` and `STT.*` commands of BASIC!

**Comment by jMarcS** [comment]
I never commented on this topic, for which I apologize. This is an essential change. It was badly needed and is gratefully received.
Released in Version 01.91, 2017/03/14.
